planner/util/path.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/ranger"
+)
+
+// AccessPath indicates the way we access a table: by using single index, or by using multiple indexes,
+// or just by using table scan.
+type AccessPath struct {
+	Index          *model.IndexInfo
+	FullIdxCols    []*expression.Column
+	FullIdxColLens []int
+	IdxCols        []*expression.Column
+	IdxColLens     []int
+	Ranges         []*ranger.Range
+	// CountAfterAccess is the row count after we apply range seek and before we use other filter to filter data.
+	CountAfterAccess float64
+	// CountAfterIndex is the row count after we apply filters on index and before we apply the table filters.
+	CountAfterIndex float64
+	AccessConds     []expression.Expression
+	EqCondCount     int
+	EqOrInCondCount int
+	IndexFilters    []expression.Expression
+	TableFilters    []expression.Expression
+	// IsTablePath indicates whether this path is table path.
+	IsTablePath bool
+	// Forced means this path is generated by `use/force index()`.
+	Forced bool
+	// PartialIndexPaths store all index access paths.
+	// If there are extra filters, store them in TableFilters.
+	PartialIndexPaths []*AccessPath
+
+	IsDNFCond bool
+
+	StoreType kv.StoreType
```

Comment:
What is the usage of it?

Reply:
It distinguishes different engine type (tikv, tiflash) for the access path.